### PR TITLE
CLI for authenticated file sharing

### PR DESCRIPTION
**Pull request by julien-c**
ping review @mfuntowicz & @thomwolf 

(I'll fix the tests for Python 2 and Python 3.5 tomorrow)

To create an account in `staging` (used by the tests): https://moon-staging.huggingface.co/join
To create an account in `production` (used by the CLI): https://huggingface.co/join

### CODE REVIEW

File: transformers/commands/user.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+class LoginCommand(BaseUserCommand):
+    def run(self):
+        print("""
+        _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|  
+        _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|        
+        _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|    
+        _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|        
+        _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|  
+
```
Comment (LysandreJik) [inline transformers/commands/user.py]: Stunning design

**Comment by LysandreJik** [comment]
Seen in person with @julien-c, really slick implementation!

**Review by LysandreJik — APPROVED** [review]

File: transformers/hf_api.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import os
+from os.path import expanduser
+
+import requests
+from requests.exceptions import HTTPError
+
+ENDPOINT = "https://huggingface.co"
+
+class S3Obj:
+    def __init__(self, filename: str, LastModified: str, ETag: str, Size: int):
```
Comment (mfuntowicz) [inline transformers/hf_api.py]: Some property are CamelCased and some are snake_case, is it on purpose ?

File: transformers/hf_api.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+class HfFolder:
+    path_token = expanduser("~/.huggingface/token")
+
+    @classmethod
+    def save_token(cls, token: str):
+        """
+        Save token, creating folder as needed.
+        """
+        os.makedirs(os.path.dirname(cls.path_token), exist_ok=True)
```
Comment (mfuntowicz) [inline transformers/hf_api.py]: What if this fails ? You let the user handle it ?

**Review by mfuntowicz — APPROVED** [review]

**Comment by stefan-it** [comment]
Can't wait to test it 😊

File: transformers/hf_api.py (hunk 2 above)
Comment (julien-c) [reply transformers/hf_api.py]: Yeah that's basically the format on the S3 response field that I just proxy through the API

Comment (julien-c) [reply transformers/hf_api.py]: Yeah that's basically the format on the S3 response field that I just proxy through the API

File: transformers/hf_api.py (hunk 3 above)
Comment (julien-c) [reply transformers/hf_api.py]: Yeah I guess. In which cases do you think it would fail?

File: transformers/commands/user.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        whoami_parser.set_defaults(func=lambda args: WhoamiCommand(args))
+        logout_parser = parser.add_parser('logout')
+        logout_parser.set_defaults(func=lambda args: LogoutCommand(args))
+        list_parser = parser.add_parser('ls')
+        list_parser.set_defaults(func=lambda args: ListObjsCommand(args))
+        # upload
+        upload_parser = parser.add_parser('upload')
+        upload_parser.add_argument('file', type=str, help='Local filepath of the file to upload.')
+        upload_parser.add_argument('--filename', type=str, default=None, help='Optional: override object filename on S3.')
+        upload_parser.set_defaults(func=lambda args: UploadCommand(args))
```
Comment (julien-c) [inline transformers/commands/user.py]: @mfuntowicz I hesitated between doing `n` subclasses of `BaseTransformersCLICommand` vs. grouping all of them in one `register_subcommand` call like here.

In the current state, `UserCommands` is not even a real `BaseTransformersCLICommand` (it doesn't have a `run()`) so I can maybe remove the reference to `BaseTransformersCLICommand `. What do you think?

File: transformers/commands/user.py (hunk 1 above)
Comment (julien-c) [reply transformers/commands/user.py]: might be a bit too wide for some terminals though 😶

File: transformers/commands/user.py (hunk 4 above)
Comment (mfuntowicz) [reply transformers/commands/user.py]: I'm fine keeping things this way. 

I want to stick to the original design, you may want to remove ```class UserCommand```, and merge it with ```class BaseUserCommand```:

```python
class BaseUserCommand(ABC, BaseTransformersCLICommand):
    def __init__(self, args):
        self.args = args
        self._api = HfApi()
``` 

And then put all the register_subcommand on the specific ```Command``` implementation. 

Keeping a common ancestor to all the commands might be interesting if we need to document / dynamically list all the commands at some point.

Comment (julien-c) [reply transformers/commands/user.py]: Up to you really! And we can always revisit later of course.

Comment (mfuntowicz) [reply transformers/commands/user.py]: Let's go this way, I'll include the change when pushing the CLI next week

File: transformers/hf_api.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        Outputs:
+            url: Read-only url for the stored file on S3.
+        """
+        urls = self.presign(token, filename=filename)
+        # streaming upload:
+        # https://2.python-requests.org/en/master/user/advanced/#streaming-uploads
+        # 
+        # Even though we presign with the correct content-type,
+        # the client still has to specify it when uploading the file.
```
Comment (julien-c) [inline transformers/hf_api.py]: A progress bar would be awesome here but didn't find a way to get one or it involved hacking into `requests`

Can probably revisit later.

**Comment by julien-c** [comment]
If anyone wants to try it out before it's merged, instructions are:
```bash
git checkout cli_upload
pip install -e .
transformers-cli login
transformers-cli upload
```

**Comment by thomwolf** [comment]
Perfect, I love it! Very slick

File: transformers/hf_api.py (hunk 5 above)
Comment (rlouf) [reply transformers/hf_api.py]: `requests` does not support streaming upload, but I think the following package adds the functionality: https://toolbelt.readthedocs.io/en/latest/uploading-data.html

**Pull request merged**